Feature/amazing feature

Abhiraj35 opened the pull request:
Hi @adityagarwal15,

I’ve completed and committed the Google OAuth2 login feature using Passport.js as part of the assigned task:
	•	Frontend: HTML, CSS, JS
	•	Backend: Node.js + Express.js
	•	Database: MongoDB

The Google login is working as expected, and user data is being stored in MongoDB.

Could you please review and merge this pull request?

Also, I had a question regarding the Phone OTP feature mentioned in the task. I’m not entirely sure if it’s meant for login/signup or for two-factor authentication (2FA).
Once clarified, would it be okay if I create a new issue for implementing the Phone OTP feature accordingly?

Looking forward to your feedback.

adityagarwal15 commented:
@Abhiraj35 can you share a video for your implementation

adityagarwal15 commented:
> Sure!I’ve recorded a short video demonstrating the Google OAuth2 login flow
> 
>  authDemo.mov

yeah good, also #211 contributors are working on to revamp this signup/login page.

adityagarwal15 commented:
> Hi @adityagarwal15,
> 
> I’ve completed and committed the Google OAuth2 login feature using Passport.js as part of the assigned task: • Frontend: HTML, CSS, JS • Backend: Node.js + Express.js • Database: MongoDB
> 
> The Google login is working as expected, and user data is being stored in MongoDB.
> 
> Could you please review and merge this pull request?
> 
> Also, I had a question regarding the Phone OTP feature mentioned in the task. I’m not entirely sure if it’s meant for login/signup or for two-factor authentication (2FA). Once clarified, would it be okay if I create a new issue for implementing the Phone OTP feature accordingly?
> 
> Looking forward to your feedback.

OTP was for signup/login only

adityagarwal15 commented:
> @Abhiraj35 @adityagarwal15 After loggin out why the profile and logout buttons are still showing?You should get a navbar with sign up and sign in options like this
> 
> <img alt="image" width="1920" height="1140" src="https://private-user-images.githubusercontent.com/147429809/473797819-3b5c6f9d-607a-4575-8408-4dea766d2c9d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQyNDY3NDgsIm5iZiI6MTc1NDI0NjQ0OCwicGF0aCI6Ii8xNDc0Mjk4MDkvNDczNzk3ODE5LTNiNWM2ZjlkLTYwN2EtNDU3NS04NDA4LTRkZWE3NjZkMmM5ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgwM1QxODQwNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWI3ZGYxOGUwMDg0NTdjY2U5YTFkNTAwNjcyNDVmODI2MjJlNWEzODhjZTdlMDRmOGRiMDY5YjBiYzIzM2U2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sQS1pwWuQvLb3DjuOXgoGsBtxCbDPoNa74Q1PqMmWwQ">

yeah I just noticed that, fix it @Abhiraj35

Abhiraj35 commented:
hey i have fixed the profile and logout button problem,
please review it .

https://github.com/user-attachments/assets/ae354b38-7113-4d00-9baf-84f37918fec6

adityagarwal15 commented:
@Abhiraj35 Great work on the Google OAuth implementation! The core functionality is solid and I appreciate you fixing the logout bug quickly. 

Before we can merge, please address these critical items:
1. Fix the cookie security settings for development vs production
2. Correct the error variable name in googleAuthController
3. Handle the case where a user signs up with Google using an email that already exists in the system

Once these are fixed, this will be ready to merge! 


I've thoroughly reviewed the Google OAuth implementation

## **What's Working Well**

1. **Complete OAuth Flow**: Proper implementation using Passport.js with Google OAuth 2.0 strategy
2. **Database Integration**: MongoDB schema updated to handle both local and Google users
3. **Session Management**: JWT tokens and passport sessions properly configured
4. **UI Integration**: Google sign-in buttons added to both login and signup pages
5. **Bug Fix**: The logout issue was identified and fixed (profile buttons showing after logout)

## 🔧 **Required Refinements Before Merge**

### **1. Security Issues (Critical)**

```javascript
// In googleAuthController - Fix this vulnerability
res.cookie('token', tkn, {
  httpOnly: true,
  secure: true,        // This should be conditional
  sameSite: 'None',    // Too permissive for development
  maxAge: 24 * 60 * 60 * 1000,
});
```

**Fix needed:**
```javascript
res.cookie('token', tkn, {
  httpOnly: true,
  secure: process.env.NODE_ENV === 'production',
  sameSite: process.env.NODE_ENV === 'production' ? 'None' : 'Lax',
  maxAge: 24 * 60 * 60 * 1000,
});
```

### **2. Error Handling Issues (Critical)**

```javascript
// In googleAuthController - Wrong variable name
console.error('Google Auth Callback Error:', err); // Should be 'error'
```

**Fix needed:**
```javascript
console.error('Google Auth Callback Error:', error);
```

### **3. Database Schema Issues (Important)**

```javascript
// In user.js model - Potential conflict
googleId: {
  type: String,
  unique: true,
  sparse: true, 
},
```

**Issue**: No validation for duplicate emails between Google and local accounts.

**Fix needed**: Add a compound index or validation logic:
```javascript
// Add this to handle email uniqueness across auth providers
userSchema.index({ email: 1 }, { unique: true });
```

### **4. Passport Configuration Issues (Important)**

```javascript
// In utils/passport.js - Logic flaw
const user = await User.findOne({ googleId: profile.id });
if (user) {
  // Update logic here
} else {
  user = await User.create({ ... }); // This will fail if email exists
}
```

**Fix needed**: Check for existing email first:
```javascript
let user = await User.findOne({ googleId: profile.id });
if (!user) {
  // Check if user exists with same email
  user = await User.findOne({ email: profile.emails[0].value });
  if (user) {
    // Link Google account to existing user
    user.googleId = profile.id;
    user.isGoogleUser = true;
    user.authProvider = 'google';
    user.isVerified = true;
    await user.save();
  } else {
    // Create new user
    user = await User.create({...});
  }
}
```

### **5. Environment Configuration (Important)**

The `.envexample` is good, but ensure production URLs are updated:
```
GOOGLE_CALLBACK_URL=http://localhost:5000/auth/google/callback
FRONTEND_URL=http://localhost:5000
```

### **6. Missing Validation (Medium Priority)**

- No rate limiting on OAuth routes
- No CSRF protection consideration
- Missing input validation for profile data

### **7. Code Style Issues (Low Priority)**

- Inconsistent spacing in route definitions
- Missing JSDoc comments for new functions
- Some unused variables in passport.js

## 📋 **Recommended Action Plan**

### **Before Merge:**
1. **Fix the critical security and error handling issues** (Items 1 & 2)
2. **Resolve the database email conflict logic** (Items 3 & 4)
3. **Test the complete flow** with existing users who sign up with Google using same email

### **Post-Merge (Can be separate PR):**
1. Add rate limiting to OAuth routes
2. Implement proper error pages for OAuth failures
3. Add comprehensive logging for OAuth events
4. Consider adding OAuth account unlinking functionality

Abhiraj35 commented:
Hi @adityagarwal15 ,

Again ,Thank you for the detailed review and feedback!

I’ve fixed all three items you mentioned:
	1.	Updated cookie security settings based on environment (development vs production).
	2.	Corrected the error variable name in googleAuthController.
	3.	Handled the case where a user signs up with Google using an email that already exists in the system.

Please let me know if there’s anything else that needs to be updated.
Looking forward to your approval and merging the PR!

adityagarwal15 commented:
@Abhiraj35 Check out the latest repo, the signup and login pages have been updated. You can now proceed with the Google and OTP authentication accordingly.

I’d recommend forking the latest repo, working on it, and then creating a new PR.

Abhiraj35 commented:
yes i have added google Oauth with updated ui 
 please check.

adityagarwal15 commented:
@Abhiraj35 will be added soon, you can checkout contributors.md for leaderboard, it will be updated soon
